libraries/custom_appbase/include/eosio/chain/exec_pri_queue.hpp

```diff
@@ -61,22 +76,47 @@ class exec_pri_queue : public boost::asio::execution_context
       should_exit_ = [](){ assert(false); return true; }; // should not be called when locking is disabled
    }
 
-   // called from appbase::application_base::exec poll_one() or run_one()
    template <typename Function>
-   void add(int priority, exec_queue q, size_t order, Function function) {
+   void add(int priority, exec_queue q, size_t order, Function&& function) {
       assert( num_read_threads_ > 0 || q != exec_queue::read_exclusive);
       prio_queue& que = priority_que(q);
-      std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(priority, order, std::move(function)));
+      std::unique_ptr<queued_handler_base> handler(new queued_handler<Function>(handler_id::unique, priority, order, std::forward<Function>(function)));
       if (lock_enabled_ || q == exec_queue::read_exclusive) { // called directly from any thread for read_exclusive
          std::lock_guard g( mtx_ );
-         que.push( std::move( handler ) );
+         que.push( handler.release() );
          if (num_waiting_)
             cond_.notify_one();
       } else {
-         que.push( std::move( handler ) );
+         que.push( handler.release() );
       }
    }
 
+   // called from appbase::application_base::exec poll_one() or run_one()
+   template <typename Function>
+   void add(handler_id id, int priority, exec_queue q, size_t order, Function&& function) {
+      assert( num_read_threads_ > 0 || q != exec_queue::read_exclusive);
+      if (id == handler_id::unique) {
+         return add(priority, q, order, std::forward<Function>(function));
+      }
+      prio_queue& que = priority_que(q);
+      std::unique_lock g( mtx_, std::defer_lock );
+      if (lock_enabled_ || q == exec_queue::read_exclusive) {
+         // called directly from any thread for read_exclusive
+         g.lock();
+      }
+      if (!que.empty()) {
+         // find the associated priority
+         auto i = std::lower_bound(que.ordered_begin(), que.ordered_end(), priority, [](const auto& h, int priority) {
```

Comment:
I thought you would use `id` to find a duplicated task.

Reply:
What happens if a non-unique-id task and a unique-id task with the same priority exist, and the non-unique-id task is found first? You would insert the unique-id task into the queue?

Reply:
This goes with the more conservative approach of only de-dup on matching id for a given `priority` and only if it is the next item to be processed (but seems like that should be if it is the last in the priority range not the first). Seems it should only not add if the last thing added with that priority is right before it in the queue.

q9 q8 q7 q6

If q9 is non-unique and q8 is something that changes state so that the non-unique is needed then a q5 should be posted. So only de-dup if q6 is non-unique.

I'm having a hard time getting the ordered iterators to work at the moment.

Reply:
Fixed https://github.com/AntelopeIO/spring/pull/802/commits/cf8d7c12d680ef00f8d15a4e4a68d0129980e740

Reply:
I have been thinking about this. Is it worthy the complexity to de-dep? Can `on_incoming_block` itself handle duplications? I am worried we might miss processing some blocks too.

Reply:
`on_incoming_block` does already handle duplicates. You can call it as often as you like. The idea here is to avoid posting to the queue. Although, we do have to take a mutex anyway, so it only avoids the post to the queue. I'll try and see if I can measure the performance difference.